Introduce SnippetToggler and New SDK Reference (#19110)
directus/directus: docs/reference/items.md

}
```

## Get Items

List all items that exist in Directus.

### Request

<SnippetToggler :choices="['REST', 'GraphQL', 'SDK']" label="API">
<template #rest>

`GET /items/:collection`

`SEARCH /items/:collection`

</template>
<template #graphql>

`POST /graphql`

```graphql
type Query {
	<collection>: [<collection>]
}
```

</template>
<template #sdk>

```js
import { createDirectus } from '@directus/sdk';
import { rest, readItems } from '@directus/sdk/rest';

> const client = createDirectus('directus_project_url').with(rest());

Reviewer: ```suggestion
const client = createDirectus('https://directus.example.com').with(rest());
```

console.log(result);
```

</template>

</SnippetToggler>

#### Query Parameters

Supports all [global query parameters](/reference/query).

### Response

Returns an [item object](#the-item-object) if a valid primary key was provided.

### Example

<SnippetToggler :choices="['REST', 'GraphQL', 'SDK']" label="API">
<template #rest>

`GET /items/articles/15`

</template>
<template #graphql>

`POST /graphql`

```graphql
type Query {
	<collection>_by_id(id: ID!): <collection>
}
```

</template>
<template #sdk>

```js
import { createDirectus } from '@directus/sdk';
import { rest, readItem } from '@directus/sdk/rest';

const client = createDirectus('https://directus.example.com').with(rest());

const result = await client.request(readItem('articles', '1'));

console.log(result);
```

</template>

</SnippetToggler>

## Get Singleton

List the singleton item in Directus.

### Request

```js
import { createDirectus } from '@directus/sdk';
import { rest, readSingleton } from '@directus/sdk/rest';
const client = createDirectus('directus_project_url').with(rest())

const result = await client.request(readSingleton('collection_name'))

console.log(result);
```

> ::: tip Info
> 
> This endpoint is only supported by the SDK.
> 
> :::

Reviewer: ```suggestion
::: tip Info

This is only intended for singleton collections.

:::
```
This is not a separate endpoint but rather the same as "Get Items" (or "Get Item") endpoint for a `singleton` collection.

Reviewer: Perhaps this should even be included in the example for either of the get items endpoints.

Reviewer 2: Addressed in https://github.com/directus/directus/pull/19110/commits/04e8583c3957751ff0d273f755811210de9aa7d1.

	}
}
```
</template>
</SnippetToggler>

---

## Delete an Item

Delete an existing item.

### Request

<SnippetToggler :choices="['REST', 'GraphQL', 'SDK']" label="API">

<template #rest>

`DELETE /items/:collection/:id`
</template>

<template #graphql>

`POST /graphql`
```graphql
type Mutation {
	delete_<collection>_item(id: ID!): delete_one
}
```
</template>
</SnippetToggler>

### Response

Empty body.

### Example

<SnippetToggler :choices="['REST', 'GraphQL', 'SDK']" label="API">

<template #rest>

`DELETE /items/articles/15`
</template>

<template #graphql>

`POST /graphql`
```graphql
mutation {
	delete_articles_item(id: 15) {
		id
	}
}
```
</template>

</SnippetToggler>

---

## Delete Multiple Items

Delete multiple existing items.

<SnippetToggler :choices="['REST', 'GraphQL', 'SDK']" label="API">

<template #rest>

`DELETE /items/:collection`
```json
// Array
[value_1, value_2, value_3]
```

```json
// Object 
{
	"field": [value_1, value_2, value_3]
}
```

```json
// Object containing query
{
	"query": {

> 		"query_type": {

Reviewer 3: @phazonoverload was working backwards from the example here and using filter docs to see what you call the generics but I dont know if got it right

}
```

</template>
<template #sdk>

```js
import { createDirectus } from '@directus/sdk';
import { rest, updateItem } from '@directus/sdk/rest';

const client = createDirectus('https://directus.example.com').with(rest());

const result = await client.request(
	updateItem('articles', '5', {
		title: 'What is Directus and how it can help you build your next app!?',
	})
);

console.log(result);
```

</template>

</SnippetToggler>

## Update Singleton

Update a singleton item

### Request

```js
import { createDirectus } from '@directus/sdk';
import { rest, updateSingleton } from '@directus/sdk/rest';
const client = createDirectus('directus_project_url').with(rest())

const result = await client.request(updateSingleton('collection_name', {
    'item_field' : 'value'
}))

console.log(result);
```

> ::: tip Info
> 
> This endpoint is only supported by the SDK.
> 
> :::

Reviewer: Same as above this not an actual endpoint but identical to updating a single item
```suggestion
```

Reviewer 2: Addressed in https://github.com/directus/directus/pull/19110/commits/488bf7050a859aa2f37193bfde7afc344cab7593

const result = await client.request(updateSingleton('about', {
    'content' : 'Founded in 2023, this website is dedicated to...'
}))

console.log(result);
```

## Update Multiple Items

Update multiple items at the same time.

### Request

<SnippetToggler :choices="['REST', 'GraphQL', 'SDK']" label="API">
<template #rest>

`PATCH /items/:collection`

```json
{
	"keys": ["id_1", "id_2"],
	"data": {
		"field": "value"
	}
}
```

</template>
<template #graphql>

`POST /graphql`

```graphql
type Mutation {
	update_<collection>_items(ids: [ID!]!, data: [update_<collection>_input]): [<collection>]
}
```

</template>
<template #sdk>

```js
import { createDirectus } from '@directus/sdk';
import { rest, updateItems } from '@directus/sdk/rest';

const client = createDirectus('directus_project_url').with(rest());

const result = await client.request(
	updateItems('collection_name', ['id_1', 'id_2'], {
		field: 'value',
	})
);

console.log(result);
```

</template>

</SnippetToggler>

#### Query Parameters

Supports all [global query parameters](/reference/query).

#### Request Body

Object containing `data` for the values to set, and either `keys` or `query` to select what items to update.

> ::: tip
> 
> Singleton If your collection is a singleton, this endpoint will act the same as the [Update an Item](#update-an-item)
> endpoint.
> 
> :::

Reviewer: ```suggestion
::: tip

If your collection is a singleton, then use the `updateSingleton` function instead.

:::
```
